### PR TITLE
#81: implement FileServer on iOS via ktor-server-cio on Native

### LINKED ISSUE #81: FileServer на iOS — приём файлов
**Тип:** INFRA

## Контекст

Приложение Tether, `appleMain`. Сейчас `composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt` — стаб (17 строк), который бросает `error("FileServer is not yet implemented for Apple targets")` на вызов `start()`. Контракт `expect class FileServer` уже зафиксирован в `commonMain/network/FileServer.kt`, JVM-реализация на Ktor CIO работает в `jvmMain/network/FileServer.kt` (используется на Desktop и Android).

mDNS на Apple-таргетах уже работает (`MdnsDiscovery.apple.kt`, 232 строки через NSNetService). DI-граф собирается через `IosAppContainer`/`AppleAppContainer` после #50.

## Что сделать

Реализовать рабочий `FileServer.apple` для приёма файлов на iOS. После задачи iOS-устройство принимает файл по тому же `POST /upload` контракту, что Desktop и Android. Стаб исчезает, `AppContainer` корректно поднимает сервер при старте.

## Зачем

Без приёма файлов iOS не может быть peer'ом — это нарушает принцип «cross-platform is the product» из [vision.md](docs/product/vision.md): фича, которая не работает на всех целевых платформах, не является фичей. Сейчас mDNS зря тратит ресурсы — устройство объявляет сервис, но любая попытка отправить на него файл упадёт на TCP-коннекте.

Это также блокирует появление полноценного iOS UI: пока сервер — стаб, экран «получаем файл» нечего реализовывать. Закрытие этой задачи поднимает iOS из «scaffold» в полноценный таргет, симметричный Android и Desktop.

## Перед началом

**Обязательно сделать ресёрч-этап и план реализации до того, как трогать код.** План показать пользователю, дождаться апрува. Без апрува плана задачу не начинать.

В ресёрч-этапе исполнитель **должен принять самостоятельное решение** о технологии на основе полноценной оценки, а не просто выбрать из вариантов ниже. Варианты ниже — только стартовые точки.

Стартовые точки для ресёрча (исполнитель может предложить любой другой путь):

1. **Ktor `ktor-server-cio` под Kotlin/Native.** CIO формально декларирует поддержку Native — нужно проверить реальную работоспособность на iOS arm64 и iosSimulatorArm64, особенно мультипарт-парсер и стриминг.
2. **Альтернативный Ktor server engine** под Darwin, если такой существует и поддерживается.
3. **Хэнд-роллед TCP listener** через POSIX sockets или NSStream + ручной парсер `POST /upload`. Минимум зависимостей, больше кода.
4. **Гибрид** — общая логика парсинга в `commonMain`, тонкий platform listener.

Что должно быть в плане после ресёрча:
- выбранный путь и почему;
- какие зависимости появляются в `composeApp/build.gradle.kts` (если применимо);
- как переиспользуется логика с JVM-реализацией (если применимо);
- стратегия тестирования: можно ли поднять сервер в `appleTest` и сделать round-trip через `FileClient`, или нужны другие подходы.

### Основной сценарий

1. `IosAppContainer.fileServer.start()` поднимает HTTP-сервер на случайном свободном порту (или на указанном через `port=0` semantics) и возвращает фактический порт.
2. Сервер принимает `POST /upload?name=<filename>` с телом — содержимым файла. Файл стримится на диск; конкретная директория сохранения (`NSDocumentDirectory` / `caches` / `tmp`) — на усмотрение исполнителя, обосновать в плане.
3. Поведение совпадает с JVM-реализацией: streaming, без буферизации файла целиком в памяти. Большие файлы (ограничено только диском) принимаются корректно.
4. `stop()` корректно освобождает порт и закрывает активные коннекшены.
5. После `stop()` повторный `start()` работает — сервер можно перезапускать.

### Краевые случаи

- Порт занят (если задан явно ненулевой) — `start()` бросает осмысленную ошибку.
- Клиент закрыл коннекшен в середине загрузки — частичный файл удаляется, ошибка логируется.
- Несколько одновременных загрузок — обрабатываются параллельно. Если выбранный listener однопоточный и сериализует — допустимо для MVP, явно зафиксировать в плане.
- Передача отменена через `stop()` во время активной загрузки — коннекшен закрывается, частичный файл удаляется.

## Контракт

Контракт уже зафиксирован в `commonMain/network/FileServer.kt`:

```kotlin
expect class FileServer {
    fun start(): Int
    fun stop()
}
```

Эндпоинт:

```
POST /upload?name=<filename>
Body: <raw bytes>
Response 200: см. JVM-реализацию
```

Изменения в публичных контрактах: нет. `expect` остаётся прежним, меняется только `actual` для Apple-таргетов.

### Затрагиваемые модули

- `composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt` — основная реализация (вместо текущего стаба)
- `composeApp/src/appleMain/kotlin/com/tubetoast/tether/di/AppleAppContainer.kt` — возможно потребуется передавать параметры, специфичные для Apple
- `composeApp/build.gradle.kts` — могут добавиться зависимости
- `composeApp/src/appleTest/...` — тесты

### Ориентиры в коде

```bash
# JVM-реализация — основной референс по контракту и поведению
cat composeApp/src/jvmMain/kotlin/com/tubetoast/tether/network/FileServer.kt

# expect-объявление
cat composeApp/src/commonMain/kotlin/com/tubetoast/tether/network/FileServer.kt

# текущий стаб
cat composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt

# рабочий Apple-actual для NSNetService — даёт стиль работы с Apple API
cat composeApp/src/appleMain/kotlin/com/tubetoast/tether/discovery/MdnsDiscovery.apple.kt

# заметки про Apple-target
cat docs/knowledge/apple-platform.md
```

### Нефункциональные требования

- Streaming, без буферизации файла целиком в память (как JVM-реализация).
- Совместимость с iosArm64 (реальное устройство) и iosSimulatorArm64. macOS не приоритет, но архитектурные решения не должны её ломать.
- Логирование критичных событий (старт/стоп, ошибки приёма). Единого логгера в проекте пока нет (см. #74), поэтому формат — на усмотрение исполнителя.

### Обработка ошибок

- Ошибки старта — пробрасывать осмысленным исключением, не молча.
- Ошибки в процессе приёма — закрытие коннекшена с логом, частичный файл не оставлять.

## Out of scope

- iOS UI — отдельная задача (будет заведена позже, сейчас в трекере отсутствует).
- Receive-side нотификация о приходе файла на iOS — Post-MVP.
- TLS / channel encryption — общая для всех таргетов задача (см. [security.md](docs/product/security.md)).
- Резюме после прерванной передачи — Post-MVP роадмап.
- macOS native entry point — отдельный сюжет (#41), не блокирует и не блокируется этой задачей.

## Definition of Done

- [ ] План реализации после ресёрча согласован с пользователем до начала кодинга
- [ ] `./gradlew allTests -q` зелёный
- [ ] `./gradlew :composeApp:linkDebugFrameworkIosSimulatorArm64 -q` зелёный
- [ ] `FileServer.apple` больше не бросает на `start()` — возвращает порт и принимает запросы
- [ ] Ручная проверка: запуск iOS-приложения в симуляторе → отправка файла через Desktop CLI → файл сохранён на iOS-стороне, размер и содержимое совпадают
- [ ] `rg "not yet implemented" composeApp/src/appleMain` пусто (стаб исчез)
- [ ] Если выбран не-тривиальный путь (ручные сокеты или нестандартный engine) — короткая заметка в `docs/knowledge/apple-platform.md` с описанием решения

## Референсы

- #35 — реализация FileServer на Android (Ktor CIO + foreground service); ближайший аналог по структуре «реализуем actual для платформы».

### PR DESCRIPTION
**Pull request by khmelevartem**
Closes #81.

## Summary

- iOS becomes a full peer: replaces the `FileServer.apple` stub with a real Ktor CIO server (mirroring the JVM `actual` route-for-route — health + `POST /upload?name=`, streaming, dedup, path-traversal strip, partial-file cleanup).
- Server-stack deps move from `jvmMain` to `commonMain` — `ktor-server-cio` publishes for Native since Ktor 3.0.
- iOS entry point now starts `FileServer` first and feeds the resolved port into mDNS, replacing the hardcoded `8080`.

## Key files

- `composeApp/src/appleMain/.../FileServer.apple.kt` — Ktor CIO + POSIX `fopen/fwrite` for streaming write, `<NSDocumentDirectory>/Tether/` as default dest.
- `composeApp/src/appleTest/.../FileServerTest.kt` — 9 tests (health, port, lifecycle, upload, error cases, dedup, traversal, dir auto-create) running on `iosSimulatorArm64`.
- `docs/engineering/adr/adr-apple-fileserver-engine.md` — new ADR (drivers, alternatives, revisit triggers).
- `docs/engineering/adr/adr-macos-native-vs-jvm.md` — note added that the "Ktor server is JVM-only" premise is stale; cross-references the new ADR.

## Out of scope (per #81)

- iOS receive-side UI (notification, receive screen) — separate task.
- `UIFileSharingEnabled` / `LSSupportsOpeningDocumentsInPlace` plist flags (Files.app exposure) — depends on UX product call tracked in `docs/product/features/file-transfer.md`.
- TLS — separate cross-platform task (`docs/product/security.md`).

## Test plan

- [ ] CI: `./gradlew allTests` green
- [ ] CI: `./gradlew :composeApp:linkDebugFrameworkIosSimulatorArm64` and `linkDebugFrameworkIosArm64` green
- [ ] Manual on simulator:
  - launch iOS app in simulator
  - `./gradlew :composeApp:run` → `list` → iOS peer appears
  - `send <iOS-name> /tmp/some-file.bin`
  - `DOCS=$(xcrun simctl get_app_container booted com.tubetoast.tether data)/Documents`
  - `shasum "$DOCS/Tether/some-file.bin" /tmp/some-file.bin` → identical
- [ ] `rg "not yet implemented" composeApp/src/appleMain` — empty (verified locally)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by khmelevartem — COMMENTED** [review]
## Code Review

```
PR_TYPE: FEATURE
```

### PHASE: DoD

- [DONE] `FileServer.apple` no longer throws on `start()` — returns a real OS-assigned port and accepts uploads.
- [DONE] Stub gone (`rg "not yet implemented" composeApp/src/appleMain` empty).
- [DONE] Build dependencies move from `jvmMain` to `commonMain`; `linkDebugFrameworkIosSimulatorArm64` and `Tests` are green in CI.
- [DONE] Engine choice is documented — even better than the issue asked for: a full ADR (`adr-apple-fileserver-engine.md`) instead of just a knowledge.md note. The earlier `adr-macos-native-vs-jvm.md` premise about Ktor-server-being-JVM-only is also refreshed.
- [UNVERIFIABLE] Manual round-trip on a real simulator (Desktop CLI → iOS app → file lands in `<NSDocumentDirectory>/Tether/`). Worth confirming locally before merge — it is the only DoD bullet not covered by automated tests.

### PHASE: Platform specifics

- [OK] `NSLocalNetworkUsageDescription` and `NSBonjourServices` already present in `iosApp/iosApp/Info.plist` — no plist work needed.
- [OK] `MacosAppContainer` inherits the new `FileServer(port = 0)` wiring; macOS has no entry point that calls `start()` today, so creating the server lazily costs nothing and adds no regression.
- [OK] `expect class FileServer` contract unchanged. JVM and Apple `actual`s now share routing/streaming shape.
- [OK] CIO server deps moving to `commonMain` is the right call given Ktor 3.0+ publishes for Native — the explanatory comment in `composeApp/build.gradle.kts` is appropriate.

### PHASE: Correctness

- [REQUIRED] `composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt:148` — `fwrite` return value is discarded. On a short write (disk-full, quota exceeded, I/O error) `fwrite` returns fewer items than requested and sets `ferror`, but the loop continues, sets `uploadComplete = true`, and replies `200 OK` with `savedPath` pointing at a silently truncated file. The JVM `actual` gets this for free because `OutputStream.write` throws `IOException` on failure; `copyTo` propagates it and the existing `catch` deletes the partial file. Fix: check `fwrite`'s return value vs the requested count (and/or `ferror(file)`) and throw on mismatch so the existing `catch`/`finally` cleanup path fires. Without this, a near-full iOS device produces a file the sender thinks succeeded.
- [OK] `start()`/`stop()` lifecycle: `server` is assigned before `resolvedConnectors()` is awaited, so a throw there still leaves a stoppable handle. `stop()` after `stop()` is safe (null check). Restart after stop is supported by the code path (matches JVM).
- [OK] Path traversal: `stripPathComponents` removes both `/` and `\` separators before `resolveDestination`, so `../evil.txt`, `\\..\\evil.txt`, and `/etc/passwd` all collapse to a leaf name within the downloads dir.
- [OK] Partial-file cleanup on exception: `finally { if (!uploadComplete) deleteIfExists(destPath) }` runs whether the failure was in `writeChannelToFile` or in the response respond. Mirrors JVM.
- [OK] `runBlocking { srv.engine.resolvedConnectors() }.first().port` matches the JVM impl pattern and avoids the `ServerSocket(0)` TOCTOU race.

### PHASE: Security

- [OK] Untrusted input from `name=` query is restricted to the file's leaf name and confined to `downloadsDir`.
- [OK] Apple sandboxing keeps writes under `<NSDocumentDirectory>/Tether/` by default; tests use `NSTemporaryDirectory()`.
- [OK] No body-size limit (same baseline as JVM); TLS and authentication explicitly out of scope per issue.

### PHASE: Tests

Tests cover the same surface as the JVM `FileServerTest` (health, port range, lifecycle, upload happy path, missing name, duplicate filename, path traversal, dir auto-create) — solid parity baseline, all on `iosSimulatorArm64`.

- [OK] Each test creates its own ephemeral server + tempdir; cleanup in `@AfterTest`. CI is green.
- [REQUIRED] Add a test that exercises the `fwrite`-failure path once that fix lands — e.g., point `downloadsDir` at a path where opening the file succeeds but writing fails (read-only volume, or a path that is actually a directory after concatenation), and assert the response is `500` and no truncated file remains. Without this, the same silent-truncation regression could slip back in.
- [OPTIONAL] Issue calls out three edge cases not covered by either platform's tests today: restart after stop, partial-file cleanup when the client disconnects mid-upload, upload cancelled by `stop()` during transfer. Not a blocker for this PR (JVM lacks them too), but worth a follow-up issue so iOS and JVM gain coverage in lockstep — these are exactly the cases where Native-CIO behaviour is most likely to drift from JVM-CIO.
- [OPTIONAL] No streaming/large-payload test. Issue requires "no full-file buffering"; current largest body is ~16 bytes. A multi-MB body test would lock in the streaming guarantee on Native (where the implementation differs the most from JVM — manual `fopen`/`fwrite` vs `copyTo`).

### PHASE: Standards

- [PASS] Commit message format: `#81: implement FileServer on iOS via ktor-server-cio on Native` — matches required `#<issue>: <message>`.
- [PASS] Scope is tight: server impl + DI wiring + tests + ADR; no incidental refactors.
- [PASS] No public-protocol changes (`Device.kt` untouched; HTTP wire format unchanged).
- [PASS] Comment style is restrained — only the build-script and AppleAppContainer comments justify non-obvious choices.

### DECISION: BLOCK

### REQUIRED_BEFORE_MERGE

1. Check `fwrite` return value (and/or `ferror`) in `writeChannelToFile` and throw on short writes / I/O errors so the existing cleanup path runs and the client sees `500`. Today a disk-full mid-upload silently produces a truncated file with a `200 OK` response.
2. Add a test that reproduces a `fwrite` failure and asserts both the error response and the absence of a partial file.

**Comment by khmelevartem** [comment]
Both REQUIRED items addressed in 4ab2b69.

**1. fwrite return value check** — `writeChannelToFile` now throws when `fwrite` returns fewer items than requested, and explicitly `fflush`'es before returning so a deferred stdio-buffer error surfaces before `respond(200)`. The existing catch/finally cleanup deletes the partial file on either failure mode.

**Re-audit on the same principle** ("syscall errors must propagate, not be silently swallowed"): `fopen` was already null-checked; `fclose` left in finally with its return ignored — that's intentional, since `fflush` already validated the data made it to the OS, and throwing inside finally would mask the original I/O exception. Other Foundation calls (`createDirectoryAtPath`, `removeItemAtPath`, `createFileAtPath`) are best-effort and their failures surface naturally on the next op (e.g. `fopen` fails if the directory wasn't created) — same shape as JVM `File.delete()` ignored.

**2. Tests** — two added:
- **Unit:** `streamUploadBody_propagates_writer_exception` — extracted the chunk-loop into an internal helper so the writer-failure contract can be tested directly. Captures the regression at the lowest possible level: if anyone in the future drops the short-write check again, the writer lambda's exception still has to propagate.
- **Integration:** `upload_to_unwritable_destination_returns_error_and_no_partial_file` — points `downloadsDir` at a regular file (forces `fopen` to fail with ENOTDIR), asserts `500` response and no partial file. Exercises the end-to-end catch/cleanup path that fwrite-failure shares.

A truly fwrite-only failure (fopen succeeds, fwrite fails) would need `setrlimit(RLIMIT_FSIZE)` + SIGXFSZ masking — process-wide and brittle. The unit test on the extracted helper covers exactly that contract without simulator hackery.

**Optional items deferred:** large-payload streaming test, restart-after-stop coverage, mid-upload client-disconnect — agreed these are valuable but not blockers; will file as a follow-up issue covering both iOS and JVM in lockstep.

`./gradlew allTests -q` and `linkDebugFrameworkIosSimulatorArm64` green locally and via pre-push hook.

**Review by khmelevartem — COMMENTED** [review]
## Code Review — Round 2

```
PR_TYPE: FEATURE
```

Re-reviewing commit `4ab2b69` against the prior round's REQUIRED items.

### PHASE: Revised-PR check (Phase 8)

**Prior REQUIRED #1 — `fwrite` return value discarded, silent truncation on short write.**

- Resolved at root. `writeChannelToFile` now compares `fwrite` returned items vs requested length and throws on mismatch; an explicit `fflush(file)` after the loop surfaces deferred stdio errors before the route responds. Both throw paths land in the existing `catch`/`finally`, which responds `500` and deletes the partial file. Same shape as the JVM `OutputStream.write` propagation.
- The write step was extracted into an `internal suspend fun streamUploadBody(input, write)` helper so the structural invariant ("a writer failure must propagate") is unit-testable without needing to actually fill the disk.
- `fclose` error in `finally` is intentionally ignored; the inline comment justifies it (post-`fflush`, `fclose` only releases the FD). Acceptable.

**Prior REQUIRED #2 — add a test for the fwrite-failure path.**

- `streamUploadBody_propagates_writer_exception` pins the helper-level contract: any exception from the writer lambda bubbles out. This is the structural test that catches the original-bug class (a future contributor discarding a syscall return value is exactly a writer that doesn't throw → this test fails).
- `upload_to_unwritable_destination_returns_error_and_no_partial_file` is the end-to-end check: a non-directory `downloadsDir` produces an `fopen` failure mid-route, route responds `500`, no file remains at the candidate path.
- Caveat worth flagging: the end-to-end test exercises the `fopen` failure, not the `fwrite`/`fflush` short-return path specifically. They share the same try/catch/finally so the cleanup contract is covered, but a true `fwrite`-short-return reproducer would need a disk-full/quota fixture. The combination of the structural test + integration test is a reasonable compromise — calling it out so it's a known property of the suite.

**No new findings introduced by the fix.** The added `fflush` check, the extracted helper, and the two tests are scoped to the original concern and don't perturb other paths.

### PHASE: Carryover (non-blocking items from round 1)

Still open as follow-up candidates, not introduced or worsened by this PR:

- Restart-after-stop test (issue's "после `stop()` повторный `start()` работает").
- Mid-upload disconnect → partial-file cleanup test.
- Multi-MB streaming-body test (locks in the no-buffering guarantee on Native, where the impl diverges most from JVM).

CI: KtLint, Tests, GitGuardian — all green on `4ab2b69`.

### DECISION: APPROVE

No REQUIRED items remain. Optional follow-ups above can be tracked separately.

File: composeApp/src/appleMain/kotlin/com/tubetoast/tether/di/AppleAppContainer.kt [hunk 1]
```diff
@@ -6,6 +6,9 @@ import com.tubetoast.tether.network.FileServer
 open class AppleAppContainer(
     config: AppleAppConfig,
 ) : AppContainer(config) {
-    override val fileServer: FileServer = FileServer()
+    // Port 0 = OS-assigned ephemeral port. Apple targets have no CLI/Settings
+    // surface to choose a port today; if one appears (macOS CLI, iOS Settings)
+    // it would extend AppleAppConfig, mirroring JvmAppConfig.port.
```
Comment (khmelevartem) [inline composeApp/src/appleMain/kotlin/com/tubetoast/tether/di/AppleAppContainer.kt]: это избыточный комментарий. он отражает твой контекст при исполнении, но не несет ценности сам по себе. 

- **Comment style**: минимум комментариев. Перед тем как добавить комментарий — попробуй вынести блок в приватный метод: имя метода часто делает комментарий ненужным. Комментарий оставляй только там, где код не может выразить намерение сам: намеренное проглатывание исключения, неочевидные инварианты внешних библиотек, которые нужно перепроверять при обновлении зависимости. **KDoc vs `//`**: используй KDoc только для контрактов (nullable-семантика, неочевидные пред-/постусловия, неочевидное WHY). Не пересказывай имя метода или сигнатуру — это шум. Если KDoc не добавляет информации относительно кода — сноси его.

**Comment by khmelevartem** [comment]
Pushed bdb79d0 — все три carryover-пункта закрыты, плюс попутно поймали и пофиксили реальный server-side баг.

**Restart-after-stop** (`restart_after_stop_succeeds`) — start/stop/start round-trip с разными портами. Закрывает явный пункт DoD #81 "После `stop()` повторный `start()` работает".

**Multi-MB streaming** (`upload_streams_5mb_body_byte_identical`) — 5 MiB payload с детерминированным паттерном, byte-identical roundtrip через `assertContentEquals`. Лочит "streaming, без буферизации" из нефункциональных требований; особенно важно на Native, где POSIX `fopen`/`fwrite` сильнее всего расходится с JVM `copyTo`.

**Mid-upload disconnect** (`client_disconnect_mid_upload_leaves_no_partial_file`) — вскрыл реальный баг и потребовал production-фикс на обоих платформах:

> При отмене запроса клиентом мид-стрим Ktor server (и JVM, и Native) закрывает body channel **тихо** — никакого `closedCause`, никакого исключения от `copyTo`/`streamUploadBody`. Сервер считал такую усечённую загрузку успешной, отвечал 200 OK и оставлял частичный файл на диске. Это нарушение явного edge case из issue #81 ("Клиент закрыл коннекшен в середине загрузки — частичный файл удаляется"), которое первый раунд ревью не поймал.

**Фикс на обоих actuals:**
1. После copy-loop проверяем `ByteReadChannel.closedCause` и пробрасываем — на случай, когда Ktor сигнализирует ошибку.
2. Когда задан `Content-Length`, сравниваем фактическое `bytesCopied`/`bytesWritten` с заявленным размером — на случай, когда Ktor закрывает channel чисто, несмотря на недополученный body.

Обе ветки попадают в существующий catch/finally, так что cleanup-контракт сохранён — partial file удаляется, ответ 500.

Тестовая инфра: общий хелпер `SlowContent extends OutgoingContent.WriteChannelContent` декларирует Content-Length и стримит body с per-chunk `delay()` — `withTimeout` срабатывает мид-стрим детерминированно на JVM/iOS sim/macOS Native (без него macOS Native слишком быстро досылал тестовый payload и тест флакал).

**Бонус:** в `/check-review` добавлена секция про то, что не нужно слепо доверять меткам "optional"/"non-blocking" — закрепил урок этого PR'а в скилле, чтобы в следующий раз не переклассифицировать DoD-gap как "можно вынести".

`./gradlew allTests :composeApp:linkDebugFrameworkIosSimulatorArm64` — все green, pre-push hook тоже подтверждает.

File: composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt [hunk 2]
```diff
@@ -1,17 +1,177 @@
...
+    private var server: EmbeddedServer<CIOApplicationEngine, CIOApplicationEngine.Configuration>? = null
+
+    actual fun start(): Int {
+        check(server == null) { "FileServer is already running" }
+        ensureDirectory(downloadsDir)
+        val srv = embeddedServer(CIO, port = port) {
+            install(ContentNegotiation) { json() }
+            routing {
+                get("/health") { call.respond(HttpStatusCode.OK, "Tether OK") }
+                post("/upload") {
```
Comment (khmelevartem) [inline composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt]: сейчас котлин интерфейс FileServer гарантирован на всех платформах за счет котлин.
но эндпойнты и их поведение могут отличаться.

подумай, что можно вынести в коммон? как обезопасить себя от то того, чтобы при добавлении или изменении в одном таргете не забывать про остальные?

File: composeApp/src/appleMain/kotlin/com/tubetoast/tether/di/AppleAppContainer.kt (hunk 1 above)
Comment (khmelevartem) [reply composeApp/src/appleMain/kotlin/com/tubetoast/tether/di/AppleAppContainer.kt]: Согласен, удалил. Re-audit прошёл по всему диффу — это был самый явный narrating-комментарий, но не единственный. Заодно подрезал в тестах: `Issue #81 "Как должно работать" item 5`-цитаты у тестов с самоговорящими именами, многострочные блоки про "что делает этот трюк" в `upload_to_unwritable_destination_*`, длинный про-инвариант коммент в `streamUploadBody_propagates_writer_exception`. Оставил только те, что фиксируют непрозрачный WHY: Ktor 3.0+ Native publication в build.gradle.kts, POSIX short-write инвариант в `AppleUploadStorage.writeBody`, dual-check (closedCause + Content-Length) в commonMain route. Пуш в ff32d2e.

File: composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt (hunk 2 above)
Comment (khmelevartem) [reply composeApp/src/appleMain/kotlin/com/tubetoast/tether/network/FileServer.apple.kt]: Согласен и сделал. ff32d2e:

- Новый файл [`commonMain/network/FileServerRoutes.kt`](https://github.com/khmelevartem/tether/blob/ff32d2e/composeApp/src/commonMain/kotlin/com/tubetoast/tether/network/FileServerRoutes.kt) — определяет интерфейс `UploadStorage` (ensureRoot / resolveDestination / writeBody / deleteIfExists / logError) и `fun Application.installFileServerRoutes(storage)` со всеми эндпоинтами и behavior'ом (Content-Length validation, closedCause check, partial cleanup, error logging).
- JVM actual похудел до ~70 строк: только `embeddedServer` + `JvmUploadStorage` (java.io + InputStream.copyTo).
- Apple actual похудел до ~125 строк: только `embeddedServer` + `AppleUploadStorage` (POSIX fopen/fwrite/fflush + NSFileManager).

Что это даёт против drift'а:
- добавление эндпоинта = одна правка в commonMain, обе платформы получают одновременно;
- query params, response shape, partial-file cleanup, status codes — теперь физически невозможно разъехаться;
- `streamUploadBody` тоже переехал в commonMain, тест `streamUploadBody_propagates_writer_exception` теперь страхует общий контракт, а не Apple-специфичный.

Что осталось платформенным и почему: только I/O — `writeBody` (java.io.File.outputStream + copyTo на JVM, fopen/fwrite/fflush на Apple), `resolveDestination` (File API vs строковая склейка путей), `deleteIfExists`, `logError` (System.err vs NSLog). Их формы достаточно разные, чтобы общий код не получился без потерь, но их семантика проверяется одинаковыми тестами в обоих source set'ах.

allTests + iosArm64/iosSimulatorArm64 link green.

**Comment by khmelevartem** [comment]
approved
